src/utils/mtev_intern.c

```diff
@@ -0,0 +1,957 @@
+    uint32_t prev;
+    while(0 != (prev = ck_pr_load_32(&ii->refcnt))) {
+      if(ck_pr_cas_32(&ii->refcnt, prev, prev+1)) break;
+    }
+    /* prev == 0, then it is being freed */
+    if(prev == 0) goto retry_fetch;
+    mtev_plock_drop_r(&pool->plock);
+  } else {
+    mtev_plock_drop_r(&pool->plock);
+    /* align len on a 4 byte boundary and add the 8 byte header */
+    size_t alen = ((len + 3) & ~3) + 8;
+
+    /* Get us a suitable new allocation in our pool */
+    ii = mtev_intern_pool_find(pool, alen);
+    if(!ii) {
+      if((void *)ibuff != (void *)lookfor) free(lookfor);
+      assert(ii); /* We don't handle failed memory allocations */
+      return mtev_intern_null;
+    }
+    /* build out our new intern object */
+    ii->poolid = pool->poolid;
+    ii->refcnt = 1;
+    ii->len = len;
+    ii->nt = nt;
+    memcpy(ii->v, buff, len);
+
+    mtev_plock_take_w(&pool->plock);
+    /* Attempt to store this in our mapping */
+    if(ck_hs_put(&pool->map, hash, ii) == false) {
+      mtev_plock_wtos(&pool->plock);
+      /* The put failed, so we need to accumulate the ii we created
+       * in our trashpile to give it back. We don't give it back here
+       * because we aren't in the right lockstate, we don't want to
+       * do that with an S or W lock held.
+       */
+      if(trash == NULL) {
+        trash = &trashpile;
+      } else {
+        trash->next = calloc(1, sizeof(*trash));
+        trash = trash->next;
+      }
+      trash->item = ii;
+
+      assert(lookfor->len == ii->len);
+      /* Get the item that is presumably there causing our put to fail. */
+      ii = ck_hs_get(&pool->map, hash, lookfor);
+      assert(ii != trash->item);
+      /* It must be here because we've been in S/W since the failed put. */
+      assert(ii);
+      /* However, someone else could have released it and refcnt -> 0 and
+       * just not completed the ck_hs_remove yet.
+       * Same as above, refcnt, but not from zero */
+      uint32_t prev;
+      while(0 != (prev = ck_pr_load_32(&ii->refcnt))) {
+        if(ck_pr_cas_32(&ii->refcnt, prev, prev+1)) break;
+      }
+      if(prev == 0) {
+        mtev_plock_stor(&pool->plock);
+        goto retry_fetch;
+      }
+      mtev_plock_drop_s(&pool->plock);
+    }
+    else {
+      mtev_plock_drop_w(&pool->plock);
+      ck_pr_inc_32(&pool->item_count);
+    }
+  }
+
+  /* This is the intern the caller is looking for */
+  rv.opaque1 = (uintptr_t)&ii->v;
+
+
+  /* cleanup */
+  if((void *)ibuff != (void *)lookfor) free(lookfor);
+  /* Free any items on our trashpile */
+  if(trashpile.item) mtev_intern_internal_release(pool, trashpile.item, mtev_false);
+  while(NULL != (trash = trashpile.next)) {
+    if(trash->item) mtev_intern_internal_release(pool, trash->item, mtev_false);
+    trashpile.next = trash->next;
+    free(trash);
+  }
+  return rv;
+}
+
+inline mtev_intern_t
+mtev_intern_pool(mtev_intern_pool_t *pool, const void *buff, size_t len) {
+  assert(len);
```

Comment:
As a general comment - do we want to use mtevAssert instead of vanilla assert throughout this? It's basically the same thing, but will assure that the assert failure will get logged if the user is using the mtev logging stuff.

Reply:
I opted for assert b/c the mtevAssert requires the logging to be initialized.

Reply:
OK. Sounds good.